### ISSUE TITLE
prompt formatting issue

### ISSUE THREAD

**Issue by RonanKMcGovern** [xfactlab/orpo#5]
Is this line correct?
```
        sample['prompt'] = [tokenizer.apply_chat_template([{'role': 'user', 'content': item[0]}], tokenize=False, add_generation_prompt=True) for item in sample['chosen']]
```

Wouldn't item[0] instead have to be pulling out the first user message?

I'm using https://huggingface.co/datasets/argilla/OpenHermesPreferences and am just pulling out the prompt as follows:
```
        sample['prompt'] = [tokenizer.apply_chat_template([{'role': 'user', 'content': item}], tokenize=False, add_generation_prompt=True) for item in sample['prompt']]
```

But maybe I'm missing why item[0] could work (when I use that, it doesn't give a tokenized prompt).

**Comment by jiwooya1000** [comment]
Yes you are correct! My mistake😅

It should be either

`sample['prompt'] = [tokenizer.apply_chat_template([item[0]], tokenize=False, add_generation_prompt=True) for item in sample['chosen']]` if the `prompt` does not explicitly exist as a separate column
or
`sample['prompt'] = [tokenizer.apply_chat_template([{'role': 'user', 'content': item}], tokenize=False, add_generation_prompt=True) for item in sample['prompt']]` otherwise.

The main code handles it automatically, but I did not integrate it to [test_orpo_trainer_demo.py](https://github.com/xfactlab/orpo/blob/main/trl/test_orpo_trainer_demo.py) as it is a simple demo code for using ORPOTrainer in TRL.

Thank you for the point! Will push the change right away👍

**Comment by RonanKMcGovern** [comment]
Thanks @jiwooya1000 . Just one further clarification.

Right now, both the user and assistant message are being formatted and included within the "chosen" field and also the "rejected" field. Is this possibly in error? Rather, should the chosen and rejected fields include only the assistant message?

I am training TinyLlama and am finding that the prompt is being repeated back to me in addition to a response thereafter:
```
<s> [INST] What is one plus one? [/INST]<s> [INST] What is one plus one? [/INST] One plus one is 2. This is because 1 plus 1 is 2.

In other words, the answer to the question "What is one plus one?" is 2.</s>
```

**Comment by jiwooya1000** [comment]
For the first part, it is intended to use the concatenated messages for both chosen and rejected fields.

Basically, the NLL loss is calculated through the full forward pass of the chosen field. With OR loss, we want to extract the log probs of the chosen/rejected responses from the assistant, which we can achieve by 1) full forward pass of chosen and rejected fields (including the user's message) and 2) masking out the log probs of the user's message. This is internally handled inside the ORPOTrainer. 

And for the second part, it did not happen for `argilla/dpo-mix-7k` dataset and `<|user|> <|assistant|>` template.
```
<|user|>
Discuss the impact of vehicle size on fuel efficiency and elaborate on how different engine types can influence this relationship.</s>
<|assistant|>
Vehicle size can significantly impact fuel efficiency. Larger vehicles like SUVs typically weigh more than smaller ones, and it takes more energy to move a heavier object than a lighter one, hence they often consume more fuel. However, this is not the only factor at play. Aerodynamics also plays a role - larger vehicles often have a larger frontal area and therefore higher air resistance, which can also decrease fuel efficiency.
```
It seems like the prompt "What is one plus one?" is being used as a system prompt for some reason. But I think this is not expected unless the dataset is actually using the user prompt as the system prompt at the same time. Could you check if the prompt is being used as a system input and if the chosen field is properly used as an input for apply_chat_template?
(e.g., `chosen = [self.tokenizer.apply_chat_template(item, tokenize=False) for item in examples['chosen']]`)

**Comment by jiwooya1000** [comment]
Can I check if I understood your situation correctly?
So, is the chunk below generated from the trained TinyLlama? I thought this chunk was from the data preprocessing function.
```
<s> [INST] What is one plus one? [/INST]<s> [INST] What is one plus one? [/INST] One plus one is 2. This is because 1 plus 1 is 2.

In other words, the answer to the question "What is one plus one?" is 2.</s>
```
If so, I recommend trying with lower LR and higher beta. Maybe you can start with an LR of 1e-5 (max) first and try a beta of 0.5. In general, smaller models require higher beta than large models (e.g., 7B). I am making the training guideline for ORPO and will upload it by this week!

**Comment by RonanKMcGovern** [comment]
Hi @jiwooya1000 , I should have been more clear. Here is the prompt:
```
<s> [INST] What is one plus one? [/INST]
```
and here is the response after fine-tuning (yes, which does erroneously repeat back the question):
```
<s> [INST] What is one plus one? [/INST]<s> [INST] What is one plus one? [/INST] One plus one is 2. This is because 1 plus 1 is 2.

In other words, the answer to the question "What is one plus one?" is 2.</s>
```
What is your intuition on using a lower LR and higher beta to address that?

**Comment by jiwooya1000** [comment]
The intuition for a low learning rate (5e-6 to 2e-5, for example) originates from the fact that a high learning rate could result in an overly huge margin between the chosen/rejected responses. And this could potentially lead to the sub-optimal state, as the model might learn the train set specific preference. Plus, even lower learning rate (e.g., 5e-7) is recommended in DPO ([Rafailov et al., 2023](https://arxiv.org/abs/2305.18290) and [Tunstall et al., 2023](https://arxiv.org/abs/2310.16944)), which I believe the reason to be similar.

For Beta, currently, Beta around 0.05 to 0.1 was the best choice for 7B models in general, and 0.5 to 1.0 was the best choice for small models. For "best choice", we determine this by seeing if the model's preference is overfitting to the train set, for example, assigning log probs of chosen responses up to 0 and -5~-10 for rejected responses eventually. Generally, larger models easily learn the preference given in the dataset compared to the small models, which I believe is intuitively understandable. For the same reason, the model easily learns the preference when the dataset has a clear pattern in chosen/rejected pairs like this example:

![image](https://github.com/xfactlab/orpo/assets/20739007/c1a0a193-b59e-4cd2-aa5b-f75dd64a63d1)

For now, I think I would have to try replicating your experiment, so could you provide the exact setting for training?

**Comment by RonanKMcGovern** [comment]
Ok, I've now run 1,000 rows of the 7k mix dataset and also openhermes preferences. In both cases, if I set the prompt as in your demo script (which results in a prompt of just `<s>`), then the results seem sensible. If I set the prompt to the actual prompt (properly tokenized), then I get repetition in the response.

Here is my trainer setup:
```
from trl import ORPOTrainer, ORPOConfig

orpo_config = ORPOConfig(
    beta=0.2, # the lambda/alpha hyperparameter in the paper/code
    # max_steps=160, # comment this out after the first time you run. This is for testing!
    save_steps=500, ### MAKE SURE TO CHECK THIS VALUE IS GOOD FOR YOUR RUN!
    logging_steps=1,
    num_train_epochs=epochs,
    output_dir=save_dir,
    evaluation_strategy="steps", #or 'epoch'
    do_eval=True,
    eval_steps=0.2,
    per_device_eval_batch_size=batch_size,
    per_device_train_batch_size=batch_size,
    gradient_accumulation_steps=grad_accum,
    log_level="debug",
    # optim="paged_adamw_8bit",
    # fp16=True, #For non-Ampere GPUs
    bf16=True, # For Ampere GPUs
    max_grad_norm=0.3,
    lr_scheduler_type="constant",
    hub_private_repo=True,
    warmup_ratio=0.03, # optional, may help stability at the start of training. Not required for simple fine-tunes.
    optim="adamw_torch", #comment out for LoRA +
    learning_rate=1e-4, #comment out for LoRA +

    max_prompt_length=512,
    max_length=1024,

    max_completion_length=1024,
    remove_unused_columns=True,
    # gradient_checkpointing=True, 
    # gradient_checkpointing_kwargs={'use_reentrant':True},

    report_to='wandb',
)

orpo_trainer = ORPOTrainer(
    model,
    args=orpo_config,
    train_dataset=train,
    eval_dataset=test,
    tokenizer=tokenizer,

    # peft_config=peft_config, #comment out if passing a peft model directly as 'model'
    callbacks=[logging_callback],  # Add custom callback here
    # neftune_noise_alpha=5 # Add in noise to embeddings to improve performance!
)
```

**Comment by RonanKMcGovern** [comment]
Based on reviewing HF docs, I'm inclined to think that actually the prompt does need to be there, but the chosen and rejected features should be limited to the responses (i.e. not include the prompt). See [here](https://github.com/huggingface/trl/issues/1484).

**Comment by jiwooya1000** [comment]
I think your point is valid here.
As I did not directly participate in building the ORPOTrainer in TRL, and that part is a bit different from our code, it seems like I missed that part🥲

[Here](https://github.com/huggingface/trl/blob/main/trl/trainer/orpo_trainer.py#L342), the preprocessing is done by concatenating the `prompt` and `chosen` field. So, one workaround is to wrap the prompts with the chat templates before passing the dataset to the Trainer. I will validate and update this part for the TRL demo code!

I will let you know as I push the new demo file tomorrow!
Thank you so much for debugging the demo code🙏

**Comment by jiwooya1000** [comment]
The test code has been updated according to the TRL ORPOTrainer document. 
Thank you for reporting the error!
Closing this issue as it is resolved, please feel free to open it whenever you have a related issue🙂
